Add new command "export_transfers" to save transfers to csv

sachaaaaa opened the pull request:
This pull request adds a new `export_transfers` command to the wallet cli that exports the transfers to a csv file.

Since `show_transfers` and `export_transfers` basically require the same data, I've added a new function `get_transfers` that returns a vector of `transfer_view` structs (new structure as well). That function does all the querying, calculations, sorting and can consume some of the common arguments.
Then, it's up to the caller function to display the data in the appropriate way (print or export as csv).

Notable difference between the csv file and the printed list of transfers:
- The csv file shows the "running balance" next the the transfer amount. This field is not affected by unconfirmed transfers (pending/failed/pool).
- To handle transfers with multiple outputs, the csv file displays a new line for any subsequent outputs after the first one,  showing only the destination address and amount sent:
```
x, <total amount>, xxxxx, xxxxxxxxxxxxxxxxx, <address1>, <amount 1>
 ,               ,      ,                  , <address2>, <amount 2>
```

Changes made to both:
- display the (shortened) destination address for incoming transfers, to easily match which address the funds were sent to. This will make it easier to match the transfer with the owner's accounts/addresses.

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6423,7 +6420,20 @@ bool simple_wallet::show_transfers(const std::vector<std::string> &args_)
-        message_writer() << (boost::format("%8.8s %6.6s %25.25s %20.20s %s %s %d %s %s%s") % "pool" % "in" % get_human_readable_timestamp(pd.m_timestamp) % print_money(pd.m_amount) % string_tools::pod_to_hex(pd.m_tx_hash) % payment_id % pd.m_subaddr_index.minor % "-" % note % double_spend_note).str();
+        
+        transfers.push_back({
+          tr("pool"),
```
Not sure writing translated data to a formatted file is a good idea.

sachaaaaa replied:
Should be ok now.
I have used `tr()` only in `show_transfers` but not in `export_transfers`

stoffu commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6423,7 +6420,20 @@ bool simple_wallet::show_transfers(const std::vector<std::string> &args_)
+          pd.m_tx_hash,
+          payment_id,
+          0,
+          {{"-", pd.m_amount}},
```
Why not do the same as in the "in" case?
```c++
      std::string destination = m_wallet->get_subaddress_as_str({m_current_subaddress_account, pd.m_subaddr_index.minor});
```

sachaaaaa replied:
Fixed.

sachaaaaa commented:
Thanks guys for the review.
I will update the diff once #4153 is merged.

moneromooo-monero commented:
It is merged now.

moneromooo-monero commented:
ping :)

moneromooo-monero commented:
If you want it in 0.13, now would be a good time to be just in time.

moneromooo-monero commented:
OK, I'll ping from time to time then.

moneromooo-monero commented:
ping :)

sachaaaaa commented:
Pong :)

Rebased on master, and all reviews addressed!
From the original diff, I have also added the "unlocked/locked" field in the transfer_view structure.
I was only able to do quick tests with "in", "out" and "block" transfers.

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -7034,7 +7034,7 @@ bool simple_wallet::show_transfers(const std::vector<std::string> &args_)
     message_writer(color, false) << formatter
       % transfer.block
-      % transfer.direction
+      % tr(transfer.direction.c_str())
```
You use tr on literals, since a tool gathers the literal to put it in a file. So tr should be moved to whatever code assigns transfer.direction.

sachaaaaa replied:
Oh ok, didn't know.
If I move the tr to whatever assigns transfer.direction, then the csv will also be using the translated string. Should I just not use tr as all?

moneromooo-monero replied:
Ah, good point. Hmm. I suppose if they both have to have the same strings, translated is probably best...

stoffu commented:
You can simply drop the last two commits.
